File: hotpot/hotpot/report/items_review_report/items_review_report.js
// Copyright (c) 2025, Bytepanda Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt

frappe.query_reports["Items Review Report"] = {
	formatter: function (value, row, column, data, default_formatter) {
        if (column.fieldname === "all_feedbacks" && value) {
            return value;
        }
        return default_formatter(value, row, column, data);
    },
	onload: function (report) {
		report.page.trigger_ready = false;
		const roles = frappe.user_roles;
		const hasAdminRole = roles.includes('Administrator');
		const custom_css = `
            /* Allow rows to expand */
            .datatable .dt-row {
                height: auto !important;
            }

            /* Allow cells to wrap multi-line content */
            .datatable .dt-cell, 
            .datatable .dt-cell__content {
                height: auto !important;
                white-space: normal !important;
                overflow: visible !important;
                line-height: 1.4em !important;
                padding-top: 4px !important;
                padding-bottom: 4px !important;
            }
        `;

        // inject only once
        if (!document.getElementById("datatable-row-height-fix")) {
            const style_element = document.createElement("style");
            style_element.id = "datatable-row-height-fix";
            style_element.innerHTML = custom_css;
            document.head.appendChild(style_element);
        }

		setTimeout(() => {
            if (report.datatable) {
                report.datatable.refresh();
            }
        }, 300);
        
		if (!hasAdminRole) {
			report.page.clear_menu();
			report.add_card_button_to_toolbar = () => { };
			report.add_chart_buttons_to_toolbar = () => { };
			report.page.remove_inner_button(__("Download Report"), __("Actions"));
			report.page.add_inner_button("Print", () => {
				let dialog = frappe.ui.get_print_settings(
					false,
					(print_settings) => report.print_report(print_settings),
					report.report_doc.letter_head,
					report.get_visible_columns()
				);
				report.add_portrait_warning(dialog);
			},)
			report.page.add_inner_button("PDF", () => {
				let dialog = frappe.ui.get_print_settings(
					false,
					(print_settings) => report.pdf_report(print_settings),
					report.report_doc.letter_head,
					report.get_visible_columns()
				);

				report.add_portrait_warning(dialog);
			},)
			report.page.add_inner_button("Export", () => {
				report.export_report();
			},)

		}

	},
	
	filters: [
		{
			fieldname: "vendor_id",
			label: __("Vendor"),
			fieldtype: "Link",
			width: "80",
			options: "Hotpot User",
			hidden: frappe.user.has_role("Hotpot Vendor") && !frappe.user.has_role("Administrator"),
			get_query: () => {
				return {
					query: "hotpot.hotpot.doctype.hotpot_user.get_all_vendor",
				};
			},
		},
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			width: "80",
			reqd: 1,
			default: frappe.datetime.add_days(frappe.datetime.get_today(), -1),
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			width: "80",
			reqd: 1,
			default: frappe.datetime.get_today(),
		},
	],
	refresh: function (report) {
		report.page.trigger_ready = false;
		const roles = frappe.user_roles;
		const hasAdminRole = roles.includes('Administrator');
		if (!hasAdminRole) {
			report.add_card_button_to_toolbar = () => { };
			report.add_chart_buttons_to_toolbar = () => { };
			report.page.remove_inner_button(__("Download Report"), __("Actions"));
			report.page.clear_menu();
			report.page.add_inner_button("Print", () => {
				let dialog = frappe.ui.get_print_settings(
					false,
					(print_settings) => report.print_report(print_settings),
					report.report_doc.letter_head,
					report.get_visible_columns()
				);
				report.add_portrait_warning(dialog);
			},)
			report.page.add_inner_button("PDF", () => {
				let dialog = frappe.ui.get_print_settings(
					false,
					(print_settings) => report.pdf_report(print_settings),
					report.report_doc.letter_head,
					report.get_visible_columns()
				);

				report.add_portrait_warning(dialog);
			},)
			report.page.add_inner_button("Export", () => {
				report.export_report();
			},)
		}
	}
}

File: hotpot/hotpot/report/items_review_report/items_review_report.py
# Copyright (c) 2025, Bytepanda Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt




import frappe, json

def execute(filters=None):
	if not filters:
		filters = {}

	start_date = filters.get("from_date")
	end_date = filters.get("to_date")
	vendor_id = filters.get("vendor_id")
	employee = filters.get("employee")

	if not start_date or not end_date:
		frappe.throw("Please select both Start Date and End Date")

	columns = [
		{"label": "Item Name", "fieldname": "item_name", "fieldtype": "Data", "width": 200},
		{"label": "Vendor", "fieldname": "vendor_id", "fieldtype": "Data", "width": 180},
		{"label": "Total Ratings", "fieldname": "ratings", "fieldtype": "Data", "width": 200},
		{"label": "Average Rating", "fieldname": "avg_rating", "fieldtype": "Data", "width": 150},
		{"label": "All Feedbacks", "fieldname": "all_feedbacks", "fieldtype": "HTML", "width": 500},
	]

	where_clause = "1=1"
	if vendor_id:
		where_clause += " AND hi.vendor_id = %(vendor_id)s"
	if employee:
		where_clause += " AND hr.employee = %(employee)s"

	frappe.db.sql("SET SESSION group_concat_max_len = 1000000")

	query = f"""
		SELECT
			COUNT(hr.rating ) AS ratings,
			hi.name AS item_id,
			hi.item_name,
			hi.vendor_id,
			COUNT(hi.name) AS total_feedback,
			ROUND(AVG(hr.rating) * 5, 1) AS avg_rating,
			GROUP_CONCAT(
				CONCAT(
					hu.full_name, ' (', hu.name, '): ',
					hr.review, ' (',
					FORMAT(hr.rating * 5, 1), ' ⭐',
					')'
				)
				SEPARATOR ' || '
			) AS all_reviews


		FROM `tabHotpot Meal Menu Items Rating` AS hr
		INNER JOIN `tabHotpot Meal Items` AS hi ON hi.name = hr.meal_item
		INNER JOIN `tabHotpot User` AS hu ON hu.name = hr.employee
		WHERE {where_clause}
		AND DATE(hr.creation) BETWEEN %(start_date)s AND %(end_date)s
		GROUP BY hi.name, hi.vendor_id
	"""

	data = frappe.db.sql(query, {"start_date": start_date, "end_date": end_date, "vendor_id": vendor_id, "employee":employee}, as_dict=True)

	result = []
	for d in data:
		stars = "⭐" * int(round(d.avg_rating or 0))

		feedbacks = []
		if d.all_reviews:
			for idx, review in enumerate(d.all_reviews.split("||"), start=1):
				review = review.strip()
				if review:  # skip empty ones
					feedbacks.append(f"<li>{review}</li>")

		feedbacks_html = "<ul style='margin:0; padding-left:16px;'>" + "".join(feedbacks) + "</ul>" if feedbacks else "No Feedback"

		result.append({
			"item_name": d.item_name,
			"vendor_id": d.vendor_id,
			 "ratings": d.ratings,
			"avg_rating": f"{d.avg_rating} ({stars})",
			"all_feedbacks": feedbacks_html
		})


	return columns, result
